nixos/modules/services/misc/matrix-conduit.nix

```diff
@@ -0,0 +1,233 @@
+            <link xlink:href="https://gitlab.com/famedly/conduit/-/blob/master/conduit-example.toml"/>
+            for details on supported values.
+            Note that database_path should not be set because the services reliance on systemd StateDir.
+        '';
+      };
+    };
+
+
+    config = mkIf cfg.enable {
+      assertions = [
+        {
+          assertion = cfg.settings.global.server_name != null;
+          message = "services.conduit.settings.global.server_name must be set.";
+        }
+        {
+          assertion = cfg.settings.global.allow_federation -> cfg.iHaveReadTheFederationWarning;
+          message = ''
+            Conduit upstream documentation contains the following warning:
+            You should not join Matrix rooms without asking the admins first.
+            We do not know whether Conduit is safe for general use yet, so you
+            should assume there is some chance that it breaks rooms permanently
+            for all participating users. We are not aware of such a bug today,
+            but we would like to do more testing.
+            You need to acknowledge this warning by setting
+            services.conduit.iHaveReadTheFederationWarning = true
+          '';
+        }
+      ];
+
+      systemd.services.conduit = {
+        description = "Conduit Matrix Server";
+        documentation = [ "https://gitlab.com/famedly/conduit/" ];
+        wantedBy = [ "multi-user.target" ];
+        environment = lib.mkMerge ([
+          { CONDUIT_CONFIG = configFile; }
+          cfg.extraEnvironment
+        ]);
+        serviceConfig = {
+          DynamicUser = true;
+          User = "conduit";
+          LockPersonality = true;
+          MemoryDenyWriteExecute = true;
+          ProtectClock = true;
+          ProtectControlGroups = true;
+          ProtectHostname = true;
+          ProtectKernelLogs = true;
+          ProtectKernelModules = true;
+          ProtectKernelTunables = true;
+          PrivateDevices = true;
+          PrivateMounts = true;
+          PrivateUsers = true;
+          RestrictAddressFamilies = [ "AF_INET" "AF_INET6" ];
+          RestrictNamespaces = true;
+          RestrictRealtime = true;
+          SystemCallArchitectures = "native";
+          SystemCallFilter = [
+            "@system-service"
+            "~@privileged"
+          ];
+          StateDirectory = "matrix-conduit";
+          ExecStart = "${cfg.package}/bin/conduit";
+          Restart = "on-failure";
+          RestartSec = 10;
+          StartLimitBurst = 5;
+        };
+      };
+      services.nginx = mkIf cfg.nginx.enable {
+        enable = mkDefault true;
+        virtualHosts.${cfg.settings.global.server_name} = {
+          enableACME = mkDefault true;
+          forceSSL = mkDefault true;
+          locations."= /.well-known/matrix/server".extraConfig =
+          let
+            server = { "m.server" = "${cfg.settings.global.server_name}:443"; };
```

Comment:
`cfg.settings.global.server_name` may already include the port, producing `:443:443`

Reply:
Also as a note: This is for nginx to redirect back to itself. Otherwise the other servers would try to reach conduit on `tcp/8448`.

Reply:
No. The server name may contain a port number, and it may be different to the port setting. The port setting only refers to the port on the local host, behind the reverse proxy. If the server has a port in the name, it will be the public one from before the reverse proxy. Specifying a port is required if no `.well-known` is used (IIRC), so forbidding it is not an option.

Reply:
I guess just asserting that the nginx vhost is disabled would be a solution. At the point where an administrator sets a specific port in there, they'll probably want to do their own thing anyway.